File: Source/Entities/RevSpeedDashBlock.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.BreadHelper.Entities;

[CustomEntity("BreadHelper/RevSpeedDashBlock")]

public class RevSpeedDashBlock : DashBlock
{
    private string breakSound;

    private bool refillDash;

    private bool refillSound;

    private bool destroyAttached;

    private string HorizontalFlip;

    private string VerticalFlip;
    public RevSpeedDashBlock(EntityData data, Vector2 offset, EntityID id) : base(data, offset, id)
	{
        breakSound = data.String("breakSound");
        refillDash = data.Bool("refillDash");
        refillSound = data.Bool("refillSound");
        destroyAttached = data.Bool("destroyAttached");
        HorizontalFlip = data.String("HorizontalFlip");
        VerticalFlip = data.String("VerticalFlip");
        OnDashCollide = RevOnDashed;
    }
    private DashCollisionResults RevOnDashed(Player player, Vector2 direction)
    {
        RevSpeedBreak(player.Center, direction);
        return DashCollisionResults.Ignore;
    }
    private void RevSpeedBreak(Vector2 from, Vector2 direction, bool playSound = true, bool playDebrisSound = true)
    {
        Audio.Play(breakSound, Position);
        for (int i = 0; (float)i < base.Width / 8f; i++)
        {
            for (int j = 0; (float)j < base.Height / 8f; j++)
            {
                base.Scene.Add(Engine.Pooler.Create<Debris>().Init(Position + new Vector2(4 + i * 8, 4 + j * 8), tileType, playDebrisSound).BlastFrom(from));
            }
        }
        Collidable = false;
        if (permanent)
        {
            RemoveAndFlagAsGone();
        }
        else
        {
            RemoveSelf();
        }
        Level level = Scene as Level;
        Player player = level.Tracker.GetEntity<Player>();
        if (player.StateMachine.State != 5 && player.StateMachine.State != 10)
        {
            player.StateMachine.State = 0;
            player.dashAttackTimer = 0f;
        }
        if(refillDash)
        {
            player.RefillDash();
            if (refillSound)
            {
                Audio.Play("event:/game/general/diamond_touch", Position);
            }
        }
        if (direction.X != 0f)
        {
            switch (HorizontalFlip) 
            {
                case "FlipNone":
                    break;
                case "FlipBoth":
                    player.Speed.X *= -1;
                    player.Speed.Y *= -1;
                    break;
                case "FlipX":
                    player.Speed.X *= -1;
                    break;
                case "FlipY":
                    player.Speed.Y *= -1;
                    break;
            }
        }
        if (direction.Y != 0f)
        {
            switch (VerticalFlip)
            {
                case "FlipNone":
                    break;
                case "FlipBoth":
                    player.Speed.X *= -1;
                    player.Speed.Y *= -1;
                    break;
                case "FlipX":
                    player.Speed.X *= -1;
                    break;
                case "FlipY":
                    player.Speed.Y *= -1;
                    break;
            }
        }

        /* random things rebound does idk what's important so I threw them here
        player.wallBoostTimer = 0f;
        player.forceMoveXTimer = 0f;
        varJumpSpeed = Speed.Y;
        varJumpTimer = 0.15f;
        AutoJump = true;
        AutoJumpTimer = 0f;
        gliderBoostTimer = 0f;
        wallSlideTimer = 1.2f;
        launched = false;
        lowFrictionStopTimer = 0.15f;
        */
        if (destroyAttached)
        {
            DestroyStaticMovers();
        }
    }

}

File: Loenn/entities/RevSpeedDashBlock.lua
local dropdownOptions = {
  ["event:/game/general/wall_break_dirt"] = "event:/game/general/wall_break_dirt",
  ["event:/game/general/wall_break_ice"] = "event:/game/general/wall_break_ice",
  ["event:/game/general/wall_break_wood"] = "event:/game/general/wall_break_wood",
  ["event:/game/general/wall_break_stone"] = "event:/game/general/wall_break_stone",
}

local dropdownOptions2 = {
  ["FlipNone"] = "FlipNone",
  ["FlipX"] = "FlipX",
  ["FlipY"] = "FlipY",
  ["FlipBoth"] = "FlipBoth",
}

local fakeTilesHelper = require("helpers.fake_tiles")

local RevSpeedDashBlock = {}

RevSpeedDashBlock.name = "BreadHelper/RevSpeedDashBlock"
RevSpeedDashBlock.depth = 0

function RevSpeedDashBlock.placements()
    return {
        name = "RevSpeedDashBlock",
        data = {
            breakSound = "event:/game/general/wall_break_stone",
            tiletype = fakeTilesHelper.getPlacementMaterial(),
            HorizontalFlip = "FlipX",
            VerticalFlip = "FlipY",
            blendin = false,
            canDash = true,
            permanent = false,
            refillDash = false,
            refillSound = true,
            destroyAttached = true,
            width = 8,
            height = 8
        }
    }
end

RevSpeedDashBlock.fieldOrder = {
	"x", "y",
	"width", "height",
	"breakSound", "tiletype",
	"HorizontalFlip", "VerticalFlip",
	"blendin", "canDash", "permanent", "destroyAttached",
	"refillDash", "refillSound"
}

function RevSpeedDashBlock.fieldInformation()
  return {
    tiletype = {
      options = fakeTilesHelper.getTilesOptions(),
      editable = false
    },
    breakSound = {
      options = dropdownOptions,
      editable = true
    },
    HorizontalFlip = {
      options = dropdownOptions2,
      editable = false
    },
    VerticalFlip = {
      options = dropdownOptions2,
      editable = false
    },
    destroyAttached = {
      default = true
    }
  }
end

RevSpeedDashBlock.sprite = fakeTilesHelper.getEntitySpriteFunction("tiletype", "blendin")

return RevSpeedDashBlock

